pilot/pkg/xds/endpoints/endpoint_builder.go

```diff
@@ -595,7 +595,29 @@ func ExtractEnvoyEndpoints(locEps []*LocalityEndpoints) []*endpoint.LocalityLbEn
 
 // buildEnvoyLbEndpoint packs the endpoint based on istio info.
 func buildEnvoyLbEndpoint(b *EndpointBuilder, e *model.IstioEndpoint, mtlsEnabled bool) *endpoint.LbEndpoint {
-	addr := util.BuildAddress(e.Address, e.EndpointPort)
+	addr := util.BuildAddress(e.Addresses[0], e.EndpointPort)
+
+	// This change can support multiple addresses for an endpoint, then there are some use cases for it, such as
```

Comment:
I am seeing very different behavior here with Istio vs non-Istio. I have 4 services, v4 only, v6 only, v4>v6, and v6>v4.

With istio:
```
/ $for i in {0..100}; do curl -s echo | grep IP=; done  | sort | uniq -c
    101 IP=127.0.0.6
/ $for i in {0..100}; do curl -s echo-v4v6 | grep IP=; done  | sort | uniq -c
    101 IP=127.0.0.6
/ $for i in {0..100}; do curl -s echo-v6v4 | grep IP=; done  | sort | uniq -c
    101 IP=127.0.0.6
/ $for i in {0..100}; do curl -s echo-v6 | grep IP=; done  | sort | uniq -c
    101 IP=127.0.0.6
```

ALL use v4

Without istio:
```
/ $for i in {0..100}; do curl -s echo-v6 | grep IP=; done | uniq | sort | uniq -c
      1 IP=fd00:10:244::b
/ $for i in {0..100}; do curl -s echo-v6 | grep IP=; done  | sort | uniq -c
    101 IP=fd00:10:244::b
/ $for i in {0..100}; do curl -s echo-v6v4 | grep IP=; done  | sort | uniq -c
    101 IP=fd00:10:244::b
/ $for i in {0..100}; do curl -s echo-v4v6 | grep IP=; done  | sort | uniq -c
    101 IP=fd00:10:244::b
/ $for i in {0..100}; do curl -s echo | grep IP=; done  | sort | uniq -c
    101 IP=10.244.0.11
```

All use v6, except the ipv4 only service.

I think its incorrect to use IPv6 for an IPv4 only and vis-versa. Kubernetes doesn't, they are not in the EndpointSlices, and DNS will not even resolve for the wrong IP family in kube-dns

Reply:
Are we sure WE should be doing happy eyeballs?

Consider this case:

```
apiVersion: v1
kind: Service
metadata:
  name: echo-only-v4-pod
spec:
  ipFamilies:
  - IPv6
  - IPv4
  ipFamilyPolicy: RequireDualStack
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: none
---
addressType: IPv4
apiVersion: discovery.k8s.io/v1
endpoints:
- addresses:
  - 10.244.0.16
  conditions:
    ready: true
    serving: true
    terminating: false
  nodeName: dual-control-plane
kind: EndpointSlice
metadata:
  labels:
    kubernetes.io/service-name: echo-only-v4-pod
  name: echo-only-v4-pod
ports:
- name: http
  port: 80
  protocol: TCP
```

This is a dualstack service with only a v4 IP.

How it works in other kubernetes networking product I have found is that the client should be doing happy eyeballs itself, and the network will just forward v4->v4 and v6->v6

Reply:
This does get trickier with SE though

Reply:
You're correct, I pointed out the same in the previous iteration of the PR, prioritizing one address family over the other isn't ideal, especially considering K8s behavior. With this PR we go from the load-balancing across all addresses via LEAST_REQUEST/RR/etc on master (which isn't ideal either) to prioritizing a single family based on the cluster's default. Ideally we should try to stick to the family the request was made with.

Keep in mind that in case we deploy on a DS on a IPv6-default cluster we would end up with IPv6 being prioritized as that would indirectly affect the ordering of addresses in IstioEndpoint. The table with your results would end up flipped.

See Istio endpoint in KinD Dual IPv6-first cluster:
```
leosarra@est-workstation:~/Desktop/istio$ go run istioctl/cmd/istioctl/main.go  x internal-debug endpointShardz | jq '."httpbin.istio-system.svc.cluster.local"'
{
  "istio-system": {
    "Shards": {
      "Kubernetes/Kubernetes": [
        {
          "Labels": {
            "app": "httpbin",
            "kubernetes.io/hostname": "kind-control-plane",
	    ...
            "version": "v1"
          },
          "Addresses": [
            "fd00:100:96::8",
            "100.96.0.8"
          ],
          ...
```
         

At the time of the previous PR I thought this was something I could address in the Envoy project by adding some knob to change the happy-eyeballs LB behavior.
But I did not consider the double-happy-eyeballs you mentioned, I was only thinking about the proxy and ignoring the client.
Due to client happy-eyeballs + client proxy accepting all connections, I feel envoy will never be able to smartly choose the initial IP family to check as the IP version of the original request could be misleading.

To be honest I don´t know if we can do better given these constraints :/

On a side note, I'm surprised of the 50% failure/success behavior you got in one scenario, I'm looking into that and trying to reproduce.

Reply:
> Keep in mind that in case we deploy on a DS on a IPv6-default cluster we would end up with IPv6 being prioritized in

How do you mark it IPv6-default in k8s?

Reply:
This got complex so started thinking it through in a doc: https://docs.google.com/document/d/1iUlgvmybCAa6QNwKA0-ju3EWGnrpqhBZHL4A0S6c9fE/edit.

Reply:
> > Keep in mind that in case we deploy on a DS on a IPv6-default cluster we would end up with IPv6 being prioritized in
> 
> How do you mark it IPv6-default in k8s?

You can use this config in KinD to deploy it.
The order of the families in the subnet will affect the order of addresses in the k8s resources.
```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  ipFamily: dual
  podSubnet: "fd00:100:96::/48,100.96.0.0/11"
  serviceSubnet: "fd00:100:64::/108,100.64.0.0/13"
```